src/ImageSharp/IO/ChunkedMemoryStream.cs

```diff
@@ -344,157 +193,203 @@ public override void Write(byte[] buffer, int offset, int count)
+        this.length += bytesWritten;
     }
 
-    /// <inheritdoc/>
-    public override void WriteByte(byte value)
+    /// <summary>
+    /// Writes the entire contents of this memory stream to another stream.
+    /// </summary>
+    /// <param name="stream">The stream to write this memory stream to.</param>
+    /// <exception cref="ArgumentNullException"><paramref name="stream"/> is <see langword="null"/>.</exception>
+    /// <exception cref="ObjectDisposedException">The current or target stream is closed.</exception>
+    public void WriteTo(Stream stream)
     {
+        Guard.NotNull(stream, nameof(stream));
         this.EnsureNotDisposed();
 
-        if (this.memoryChunk is null)
+        this.Position = 0;
+
+        long remaining = this.length - this.position;
+        if (remaining <= 0)
         {
-            this.memoryChunk = this.AllocateMemoryChunk();
-            this.writeChunk = this.memoryChunk;
-            this.writeOffset = 0;
+            // Already at the end of the stream, nothing to read
+            return;
         }
 
-        Guard.NotNull(this.writeChunk);
+        int bytesToRead = (int)remaining;
+        int bytesRead = 0;
+        while (bytesToRead != 0 && this.currentChunk != this.memoryChunkBuffer.Length)
+        {
+            bool moveToNextChunk = false;
+            MemoryChunk chunk = this.memoryChunkBuffer[this.currentChunk];
+            int n = bytesToRead;
+            int remainingBytesInCurrentChunk = chunk.Length - this.currentChunkIndex;
+            if (n >= remainingBytesInCurrentChunk)
+            {
+                n = remainingBytesInCurrentChunk;
+                moveToNextChunk = true;
+            }
 
-        IMemoryOwner<byte> chunkBuffer = this.writeChunk.Buffer;
-        int chunkSize = this.writeChunk.Length;
+            // Read n bytes from the current chunk
+            stream.Write(chunk.Buffer.Memory.Span.Slice(this.currentChunkIndex, n));
+            bytesToRead -= n;
+            bytesRead += n;
 
-        if (this.writeOffset == chunkSize)
-        {
-            // Allocate a new chunk if the current one is full
-            this.writeChunk.Next = this.AllocateMemoryChunk();
-            this.writeChunk = this.writeChunk.Next;
-            this.writeOffset = 0;
-            chunkBuffer = this.writeChunk.Buffer;
+            if (moveToNextChunk)
+            {
+                this.currentChunkIndex = 0;
+                this.currentChunk++;
+            }
+            else
+            {
+                this.currentChunkIndex += n;
+            }
         }
 
-        chunkBuffer.GetSpan()[this.writeOffset++] = value;
+        this.position += bytesRead;
     }
 
     /// <summary>
-    /// Copy entire buffer into an array.
+    /// Writes the stream contents to a byte array, regardless of the <see cref="Position"/> property.
     /// </summary>
-    /// <returns>The <see cref="T:byte[]"/>.</returns>
+    /// <returns>A new <see cref="T:byte[]"/>.</returns>
```

Comment:
Note that this is very inefficient; I would suggest trying to deprecate this kind of API - especially if we *can* use ROS, for example;

```
[Obsolete("prefer " + nameof(AsReadOnlySequence)] public byte[] ToArray() => AsReadOnlySequence().ToArray();
public ReadOnlySequence<byte> AsReadOnlySequence() => /* magic happens */
```

Reply:
Unfortunately, we need this to when reading XMP data for the V3 build however I wish to rewrite the XMPProfile type for V4 to avoid passing arrays around. Once this is merged to the V3 branch I'll upstream and make additional changes.